### PR TITLE
fix(engine): don't NPE when timing out job that doesn't exist anymore

### LINKED ISSUE #13123: NPE when processing Job.TIME_OUT command
**Describe the bug**

Currently, the `JobTimeOutProcessor` assumes that the job it is supposed to time out is in the state. This is wrong, as it's possible to have the following sequence on the log: Job.COMPLETE, Job.TIME_OUT. The first command will remove the job from the state, and when processing the time out command, the state will not return any job.

**To Reproduce**

New test in `JobTimeOutTest`:

```java
@Test
  public void shouldRejectIfJobDoesNotExist() {
    // given
    final var jobKey = ENGINE.createJob(jobType, PROCESS_ID).getKey();
    final var job = ENGINE.getProcessingState().getJobState().getJob(jobKey);
    final var partitionId = Protocol.decodePartitionId(jobKey);

    // when
    ENGINE.pauseProcessing(partitionId);
    ENGINE.writeRecords(
        RecordToWrite.command().job(JobIntent.COMPLETE, job),
        RecordToWrite.command().job(JobIntent.TIME_OUT, job));
    ENGINE.resumeProcessing(partitionId);
    Awaitility.await("until everything processed").until(ENGINE::hasReachedEnd);

    // then activated again
    final List<Record<JobRecordValue>> jobEvents =
        jobRecords()
            .withRecordKey(jobKey)
            .withIntent(JobIntent.TIME_OUT)
            .withRecordType(RecordType.COMMAND_REJECTION)
            .limit(1)
            .toList();
    assertThat(jobEvents).isNotEmpty();
  }
```

**Expected behavior**

The command is rejected when the job to time out does not exist.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.NullPointerException: Cannot invoke "io.camunda.zeebe.protocol.impl.record.value.job.JobRecord.getDeadline()" because "job" is null
	at io.camunda.zeebe.engine.processing.job.JobTimeOutProcessor.processRecord(JobTimeOutProcessor.java:53) ~[zeebe-workflow-engine-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.engine.Engine.process(Engine.java:143) ~[zeebe-workflow-engine-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.batchProcessing(ProcessingStateMachine.java:347) ~[zeebe-stream-platform-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.lambda$processCommand$2(ProcessingStateMachine.java:268) ~[zeebe-stream-platform-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[zeebe-db-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.processCommand(ProcessingStateMachine.java:268) ~[zeebe-stream-platform-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.tryToReadNextRecord(ProcessingStateMachine.java:227) ~[zeebe-stream-platform-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.readNextRecord(ProcessingStateMachine.java:203) ~[zeebe-stream-platform-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:92) [zeebe-scheduler-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) [zeebe-scheduler-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) [zeebe-scheduler-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) [zeebe-scheduler-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) [zeebe-scheduler-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:203) [zeebe-scheduler-8.3.0-SNAPSHOT.jar:8.3.0-SNAPSHOT]
```

</p>
</details>

### PR DESCRIPTION
**Pull request by lenaschoenburg**
Delays the lookup of the job deadline until we are sure that the state is `ACTIVATED`. Otherwise, the state could be `NOT_FOUND` and looking up a deadline would result in an NPE.

Closes #13123

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobTimeOutTest.java [hunk 1]
```diff
@@ -170,6 +174,32 @@ public void shouldHaveNoSourceRecordPositionOnTimeOut() {
...
+    Awaitility.await("until everything processed").until(ENGINE::hasReachedEnd);
+
+    // then activated again
+    final List<Record<JobRecordValue>> jobEvents =
+        jobRecords()
+            .withRecordKey(jobKey)
+            .withIntent(JobIntent.TIME_OUT)
+            .withRecordType(RecordType.COMMAND_REJECTION)
+            .limit(1)
+            .toList();
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobTimeOutTest.java:199]: 🔧 No need to call `.toList()`
🔧 `jobEvents` can be inlined into assertion

**Review by korthout — APPROVED** [review]
Nice work @oleschoenburg! 👏 

👍 Simple solution
👍 I like the test case

👍 LGTM

**Review by koevskinikola — APPROVED** [review]
:+1: Nice one, pretty clean.

**Comment by koevskinikola** [comment]
Since the bug that this PR is fixing is causing some instability in our CI, I'm merging it without the optional suggestions raised by @korthout.

bors merge

**Comment by korthout** [comment]
bors retry

**Comment by korthout** [comment]
Gonna give it one more try:

bors merge

**Pull request merged**